### PR TITLE
Add access to register custom field components

### PR DESCRIPTION
**Pull request by inrumi**
Signed-off-by: Johnn Mis <v-jmis@expediagroup.com>

## Hey, I just made a Pull Request!

Based on this #5134 I've just followed what was suggested importing the scaffolder api and adding some fields/methods to keep track of the custom fields also looked on the implementation of `OwnerPicker` & `RepoUrlPicker`.

Implementation look like this:
```jsx
// ./app/App.tsx
// const AutocompleteField: Field = ({ schema }) => {};

const api = useApi(scaffolderApiRef);
api.registerField('autocomplete', AutocompleteField);
```

Example of custom field component looks like this:
image

Template like this:
image

This requires to update the documentation and add tests? (I just think that was a very straightforward  implementation and didn't added any)

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by regicsolutions** [comment]
This is awesome!

File: plugins/scaffolder/src/api.ts [hunk 1]
```diff
@@ -235,4 +241,15 @@ export class ScaffolderClient implements ScaffolderApi {
     const response = await fetch(`${baseUrl}/v2/actions`);
     return await response.json();
   }
+
+  registerField(name: string, field: Field): void {
```
Comment (Rugvip) [inline plugins/scaffolder/src/api.ts:245]: Not feeling that an API is best suited for this kind of component registration tbh. It may be that either a dedicated type of component extension (`createScaffolderFormExtension`) or just a prop somewhere is what we should be going for instead, but not too sure tbh.

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx [hunk 2]
```diff
@@ -145,6 +145,11 @@ export const TemplatePage = () => {
   const { templateName } = useParams();
   const navigate = useNavigate();
   const rootLink = useRouteRef(rootRouteRef);
+  const fields = {
+    RepoUrlPicker,
+    OwnerPicker,
+    ...scaffolderApi.getRegisteredFields(),
```
Comment (Rugvip) [inline plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:151]: This isn't quite enough to make the fields fully customizable, which we'd want. The missing piece is the `createValidator` function above, where you can see a check for `'RepoUrlPicker'` specifically. As part of this solution we'd need to generalize that and allow these custom components to provide their own validation logic.

**Review by Rugvip — COMMENTED** [review]
Thanks! This is a bit of a tricky topic design wise. Adding some initial comments, let me know what you think

File: plugins/scaffolder/src/api.ts (hunk 1 above)
Comment (inrumi) [reply plugins/scaffolder/src/api.ts:245]: This makes sense to me since this custom component logic isn't related to the API but I can't understand _the how_ in creating a component extension or a prop, right now there's nothing to access a component outside of scaffolder besides the API/Client and `ScaffolderPage`.

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx (hunk 2 above)
Comment (inrumi) [reply plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:151]: Ohh I see I missed that!

**Comment by defaultbr** [comment]
Owww, awesome, this will help us a lot to not bloat the core code.

@inrumi  is possible to add inside the registred field the formstate? 

If we can pass fields or at least previous fields input that, will be possible to add complex logic to the custom field based on the previous input

File: plugins/scaffolder/src/api.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/scaffolder/src/api.ts:245]: I can help out with getting it in place if we go down the route of component extensions, but kinda wanna settle if it's the API we want first.

It could look something like this:

```tsx
import { createScaffolderFormExtension } from '@backstage/plugin-scaffolder';

export const AutocompleteField = autocompletePlugin.provide(createScaffolderFormExtension({
  name: 'AutocompleteField',
  component: {
    async: () => import('./components/AutocompleteField').then(m => m.AutocompleteField)
  },
  validate: (...) => /* custom validation logic */,
})
```

To be used in the app like this:

```tsx
<FlatRoutes>
  ...
  <ScaffolderPage>
    <AutocompleteField />
  </ScaffolderPage>
</FlatRoutes>
```

Or maybe a but more explicit?

```tsx
<FlatRoutes>
  ...
  <ScaffolderPage>
    <ScaffolderPage.FormFields>
      <AutocompleteField />
    </ScaffolderPage.FormFields>
  </ScaffolderPage>
</FlatRoutes>
```

Thoughts on ^ @backstage/maintainers ?

Comment (inrumi) [reply plugins/scaffolder/src/api.ts:245]: uhm I think this is not what I'm trying to get done, I as a user of scaffolder I do not have access to modify the ScaffolderPage component, what I want is a way to register my custom components to rjsf this way any template using a custom field will be rendered by the custom field, like in the documentation from rjsf.

https://react-jsonschema-form.readthedocs.io/en/docs/advanced-customization/custom-widgets-fields/#custom-field-components

TLDR: I want to add my components in `packages/app`

Comment (Rugvip) [reply plugins/scaffolder/src/api.ts:245]: It totally is. You wouldn't be modifying the `ScaffolderPage` component, just supplying some extra children to it in the app, which are then interpreted as components to add as rjsf fields.

Comment (inrumi) [reply plugins/scaffolder/src/api.ts:245]: you are right, I can see it now! 👁👁, your approach looks more advanced and decoupled from my initial implementation, gonna give it a try, thanks!

Comment (benjdlambert) [reply plugins/scaffolder/src/api.ts:245]: I like the more explicit approach and wrapping them in the `ScaffolderPage.FormFields` section.

Gives more flexibility to the JSX API too in case we start including other extension points (Which is always most probably the case).

Comment (inrumi) [reply plugins/scaffolder/src/api.ts:245]: @Rugvip Well I've been trying to wrap my head around converting this into an extension but I haven't found any documentation on how to implement it, I found [this](https://backstage.io/docs/plugins/composability#extensions-from-a-plugins-point-of-view) but is about how to use it, is there any reference I can look at?

**Comment by freben** [comment]
Hi! The core team talked among ourselves and thought that this is a really interesting use case for leveraging the composability system as mentioned above. If you don't mind, we would think it interesting to explore this a bit ourselves as a group to make sure that the system holds and fits well. It'd be one of the first significant extensions apart from the routing itself and we may find things along the way that make us want to make the system itself even better.

We hope to get onto that some time next week. Is it OK for you to hold off a little? We'll update you as we go, of course.

**Comment by inrumi** [comment]
> Hi! The core team talked among ourselves and thought that this is a really interesting use case for leveraging the composability system as mentioned above. If you don't mind, we would think it interesting to explore this a bit ourselves as a group to make sure that the system holds and fits well. It'd be one of the first significant extensions apart from the routing itself and we may find things along the way that make us want to make the system itself even better.
> 
> We hope to get onto that some time next week. Is it OK for you to hold off a little? We'll update you as we go, of course.

Sure, it is ok for me!, if there's anything I can help just reach me out please.

**Comment by guillermomanzo** [comment]
Any update on this PR review? Gotta a "couple" of customers in our company that are in need of this and we are hoping to leverage the feature as soon as it gets released :)

**Comment by freben** [comment]
We've been pretty swamped this week, haven't made progress in actually implementing the thing yet.

**Comment by pamelin** [comment]
Is this something what can be merged now and improved on by you later when you'll get time to get to this issue?

**Comment by pamelin** [comment]
Tank you for the update @benjdlambert 

Does it mean this PR will never get merged?

**Comment by Rugvip** [comment]
@pamelin Yeah this does end up being superseded by #5677

**Comment by pamelin** [comment]
@Rugvip Cool thank you for the confirmation

**Comment by benjdlambert** [comment]
Think we can close this now given #5677 was merged.
